src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -130,6 +131,131 @@ internal ProcessOutputObject(object data, MinishellStream stream)
         }
     }
 
+    #nullable enable
+    /// <summary>
+    /// This exception is used by the NativeCommandProcessor to indicate an error
+    /// when a native command retuns a non-zero exit code.
+    /// </summary>
+    [Serializable]
+    public sealed class NativeCommandExitException : RuntimeException
+    {
+        // NOTE:
+        // When implementing the native error action preference integration,
+        // reusing ApplicationFailedException was rejected.
+        // Instead of reusing a type already used in another scenario
+        // it was decided instead to use a fresh type to avoid conflating the two scenarios:
+        // * ApplicationFailedException: PowerShell was not able to complete execution of the application.
+        // * NativeCommandExitException: the application completed execution but returned a non-zero exit code.
+
+        #region Constructors
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NativeCommandExitException"/> class with information on the native
+        /// command, a specified error message and a specified error ID.
+        /// </summary>
+        /// <param name="path">The full path of the native command.</param>
+        /// <param name="exitCode">The exit code returned by the native command.</param>
+        /// <param name="processId">The process ID of the process before it ended.</param>
+        /// <param name="message">The error message.</param>
+        /// <param name="errorId">The PowerShell runtime error ID.</param>
+        internal NativeCommandExitException(string path, int exitCode, int processId, string message, string errorId)
+            : base(message)
+        {
+            SetErrorId(errorId);
+            SetErrorCategory(ErrorCategory.NotSpecified);
+            Path = path;
+            ExitCode = exitCode;
+            ProcessId = processId;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NativeCommandExitException"/> class.
+        /// </summary>
+        public NativeCommandExitException() : base()
```

Comment:
This addresses docs recommendations.

Reply:
This exception should only be thrown from PowerShell, and a user should not need to create an instance of such exception in any scenario. So, there is no need to have those public constructors.

Reply:
This follows [this](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1032#rule-description) recommendation from .NET

Reply:
That recommendation is for exception types that are supposed to be consumed by the library user. In our case, I don't think we are expecting our user to create such an exception, do we?

Reply:
> That recommendation is for exception types that are supposed to be consumed by the library user

That's the part I was trying to verify, but couldn't find any explicit explanation.

I'll take the constructors out since it's easier to add them in later than vice-versa.

Reply:
>For example, the constructor that has the signature NewException(string, Exception) is used to create exceptions that are caused by other exceptions. Without this constructor, you can't create and throw an instance of your custom exception that contains an inner (nested) exception, which is what managed code should do in such a situation.

The example indicates that this rule is for exception types that are to be consumed by library users.
In this case, we don't expect any user to create such an exception, so no public constructor is needed.

Reply:
Just above that though they say:

> Failure to provide the full set of constructors can make it difficult to correctly handle exceptions

The use of "handle" I found confusing, since I think of handling an exception as catching it.